engines/gnap/gamesys.cpp

```diff
@@ -0,0 +1,1278 @@
+		gfxItem->_prevFrame._soundId = -1;
+		gfxItem->_prevFrame._unusedVal = -1;
+		int totalDuration = duration;
+		if ((seqItem->_flags & 4) && totalDuration > 0) {
+			gfxItem->_prevFrame._duration = 1;
+			if (gfxItem->_delayTicks <= totalDuration)
+				gfxItem->_delayTicks = 0;
+			else
+				gfxItem->_delayTicks -= totalDuration + 1;
+			gfxItem->_updFlag = false;
+		} else if (gfxItem->_delayTicks <= totalDuration) {
+			int j;
+			totalDuration -= gfxItem->_delayTicks;
+			gfxItem->_delayTicks = 0;
+			for (j = gfxItem->_currFrameNum; j < animation->_framesCount && animation->frames[j]._duration <= totalDuration; ++j) {
+				if (animation->frames[j]._soundId != -1)
+					_soundIds.push_back((gfxItem->_sequenceId & 0xFFFF0000) | animation->frames[j]._soundId);
+				totalDuration -= animation->frames[j]._duration;
+			}
+			if (animation->_framesCount > j)
+				gfxItem->_currFrame = animation->frames[j++];
+			else
+				gfxItem->_currFrame = animation->frames[j - 1];
+			if (gfxItem->_currFrame._spriteId != -1 && (seqItem->_x != 0 || seqItem->_y != 0))
+				gfxItem->_currFrame._rect.translate(seqItem->_x, seqItem->_y);
+			// Update sprite scaling
+			if ((seqItem->_flags & 1) && gfxItem->_currFrame._rect.bottom >= _backgroundImageValue1 && gfxItem->_currFrame._rect.bottom <= _backgroundImageValue3) {
+				int scaleValue = _backgroundImageValue2	+ (gfxItem->_currFrame._rect.bottom - _backgroundImageValue1) *
+					(_backgroundImageValue4 - _backgroundImageValue2) /
+					(_backgroundImageValue3 - _backgroundImageValue1);
+				gfxItem->_currFrame._rect.top = gfxItem->_currFrame._rect.bottom - scaleValue * (gfxItem->_currFrame._rect.bottom - gfxItem->_currFrame._rect.top) / 1000;
+				gfxItem->_currFrame._rect.right = scaleValue * (gfxItem->_currFrame._rect.right - gfxItem->_currFrame._rect.left) / 1000 + gfxItem->_currFrame._rect.left;
+				gfxItem->_currFrame._isScaled = true;
+			}
+			gfxItem->_currFrame._duration -= totalDuration;
+			if (gfxItem->_currFrame._soundId != -1)
+				_soundIds.push_back((gfxItem->_sequenceId & 0xFFFF0000) | gfxItem->_currFrame._soundId);
+			gfxItem->_currFrameNum = j;
+			gfxItem->_updFlag = true;
+		} else {
+			gfxItem->_delayTicks -= totalDuration + 1;
+			gfxItem->_updFlag = false;
+		}
+	}
+
+	for (int k = 0; k < kMaxAnimations; ++k) {
+		if (_animations[k]._sequenceId != -1 && _animations[k]._sequenceId == seqItem->_sequenceId && _animations[k]._id == seqItem->_id) {
+			_animations[k]._status = 1;
+			break;
+		}
+	}
+}
+
+void GameSys::seqRemoveGfx(int sequenceId, int id) {
+	int gfxIndex;
+	if (seqLocateGfx(sequenceId, id, &gfxIndex)) {
+		GfxItem *gfxItem = &_gfxItems[gfxIndex];
+		while (gfxIndex < _gfxItemsCount && gfxItem->_sequenceId == sequenceId && gfxItem->_id == id) {
+			if (gfxItem->_prevFrame._spriteId == -1) {
+				--_gfxItemsCount;
+				if (gfxIndex != _gfxItemsCount)
+					memcpy(&_gfxItems[gfxIndex], &_gfxItems[gfxIndex + 1], sizeof(GfxItem) * (_gfxItemsCount - gfxIndex));
```

Comment:
Valgrind complains about overlapping memory areas to memcpy(), which is undefined. We should use memmove() instead.